synfig-studio/ui/synfig-catalog.xml

```diff
@@ -2,14 +2,18 @@
 <glade-catalog name="synfigstudio" depends="gtk+" version="0.2" domain="synfigstudio">
 
   <glade-widget-classes>
+    <glade-widget-class name="gtkmm__CustomObject_widget_gammapattern" title="Gamma Pattern" />
     <glade-widget-class name="gtkmm__CustomObject_widget_vector" generic-name="vector" title="Vector" parent="GtkEntry" toplevel="false" default-width="100" />
     <glade-widget-class name="gtkmm__CustomObject_widget_time" since="0.2" generic-name="time" title="Time" parent="GtkEntry" toplevel="false" default-width="100" />
+    <glade-widget-class name="gtkmm__CustomObject_widget_link" title="Link" />
```

Comment:
you should set the "parent" attribute as GtkToggleButton. It would make it appear better on glade ;)

Reply:
I implemented Widget_Link as a custom widget but if you open glade file you'll see that I inserted the standard GtkToggleButton. Also, in canvas_properties.cpp file I manually implemented button image and tooltips.

But if you want I can delete mine and instead call the custom widget :) But reimplementing Glib::property for Widget_Link's sake is too minimal don't you think?

Reply:
Do you mean this line?
```xml
 <object class="GtkToggleButton" id="toggle_wh_r"> 
```

Well, you could use 
```
 <object class="gtkmm__CustomObject_widget_link" id="toggle_wh_r"> 
```
This class is now available by you catalog, so it appears in glade widget list. And, if you set the parent attribute in catalog, it would appear as a toggle button.

As Widget_Link can (is?) used elsewhere, programmer should always care about changing the image and tooltip (repeating code…).

You don't have to reimplement Glib::property, just create it. A bit ugly though, but…

Reply:
Yup

Alright

Reply:
Check this commit 0e6f409c8e6e6052dcdb452e13d2c09dcb39d69d
Hope I did the Glib::Property right like you asked

Reply:
> you should set the "parent" attribute as GtkToggleButton. It would make it appear better on glade ;)

Didn't work

Still got the same output like last time
![Screenshot from 2020-07-23 14-40-56](https://user-images.githubusercontent.com/24214509/88269991-96a88580-ccf2-11ea-8151-77c6126eb438.png)

Reply:
Oh I think you didn't XD
You'd replace the tooltips member properties with Glibmm instrospectable ones.

The `property_mine_` from my example would be actually something like `property_tooltip_link_` and `property_tooltip_unlink_` .
In constructor initialization list:
```c++
property_tooltip_link_(*this, "tooltip_link", "Link");
property_tooltip_unlink_(*this, "tooltip_unlink", "Unlink");
```
(tooltip_active and tooltip_inactive as I think it is right now)

then you get (and set) the string via the proxy `property_tooltip_link()`:
```c++
set_tooltip(property_tooltip_link())
```

Reply:
I don't really understand how we are gonna send the custom tooltip_text to wrap_new which takes in the default constructor without arguments because now `tooltip_active` and `tooltip_inactive` strings are empty.

whereas for this previous code that we already had:
https://github.com/synfig/synfig/blob/8b93aa1c2baf37ca687010c617db988632df79fb/synfig-studio/src/gui/renddesc.cpp#L597-L598
We know those strings are custom initialized for each widget

Now in this PR we just have:
https://github.com/Keyikedalube/synfig/blob/d1ebfd83a56d567a20758fdd17cdcfae6185ad69/synfig-studio/src/gui/dialogs/canvasproperties.cpp#L306-L307

That get_widget calls the default constructor without string arguments:
https://github.com/Keyikedalube/synfig/blob/d1ebfd83a56d567a20758fdd17cdcfae6185ad69/synfig-studio/src/gui/widgets/widget_link.cpp#L158-L178

The strings are not custom initialized.

A total noob question but I haven't gotten into practice with deriving custom widgets. This is new.

However, if your ultimate design plan is to set the tooltip manually from Widget_Link implementation class
I think simply setting property tooltip_text inside the glade file would suffice because that works too

BTW, can you show me a sample project similar to this so I can familiarize how this is done?

Reply:
Working on making the changes :)

Also, I forgot that I'm supposed to use get_widget_derived()

A bit of roadblock but all is fine and clear now

Reply:
Made the changes to CanvasProperties like this:
```c++
...
canvas_properties->get_widget_derived("toggle_wh_r" , toggle_wh_r , "Link width and height"  , "Unlink width and height");
canvas_properties->get_widget_derived("toggle_res_r", toggle_res_r, "Link x and y resolution", "Unlink x and y resolution");
...
```

Also corrected the property_link part like this:
```c++
...
Widget_Link::Widget_Link():
	Glib::ObjectBase("widget_link"),
	property_tooltip_active_(*this, "tooltip_active", ""),
	property_tooltip_inactive_(*this, "tooltip_inactive", "")
{
	init();
}

Widget_Link::Widget_Link(const std::string &tlt_inactive, const std::string &tlt_active):
	Glib::ObjectBase("widget_link"),
	property_tooltip_active_(*this, "tooltip_active", ""),
	property_tooltip_inactive_(*this, "tooltip_inactive", "")
{
	init(tlt_inactive, tlt_active);
}

Widget_Link::Widget_Link(BaseObjectType *cobject):
	Glib::ObjectBase("widget_link"),
	Gtk::ToggleButton(cobject),
	property_tooltip_active_(*this, "tooltip_active", ""),
	property_tooltip_inactive_(*this, "tooltip_inactive", "")
{
	init();
}

Widget_Link::Widget_Link(BaseObjectType *cobject, const std::string &tlt_inactive, const std::string &tlt_active):
	Glib::ObjectBase("widget_link"),
	Gtk::ToggleButton(cobject),
	property_tooltip_active_(*this, "tooltip_active", ""),
	property_tooltip_inactive_(*this, "tooltip_inactive", "")
{
	init(tlt_inactive, tlt_active);
}
...
```

Got this error:
![Screenshot from 2020-07-24 09-52-31](https://user-images.githubusercontent.com/24214509/88360745-7af5bb80-cd94-11ea-866f-77fdca80db0a.png)

Deriving widgets is new to me so I'm finding your method a bit complicated :(

I have an alternative solution though:
Manually set the tooltip for toggle_wh_r as "width" and toggle_res_r as "resolution" and during the construction of Wdget_Link do string comparison and ultimately set the Link/Unlink phrase based on get_active() status. Of course, the string (four to be exact) has to be stored as constants inside the Widget_Link class.
Simple and easy

I'm proceeding with my suggestion

Reply:
It'

> Deriving widgets is new to me so I'm finding your method a bit complicated :(

I can't say I master it too. :/
Take a look at this example I made for you: https://gist.github.com/rodolforg/802222eb41287da744646529615d91e2

Interesting enough, you can inspect and change the widget properties via GtkInspector via Ctrl+Shift+D or Ctrl+Shift+I ;)
(make sure you have gtkinspector installed :P )

Reply:
Custom alternative solution implemented here commit f3bf66278432e7db35dbe8db23d879c8084ebeda

It works

Reply:
Your glib::property stuff is also done in completion. All you have to do is call that `ref_builder->get_custom_widgets(...)` from canvas_properties.cpp file for `toggle_wh_r` and `toggle_res_r`

Reply:
But did you understand the purpose?
It's about the UI designer to provide the tooltips him(her)self via Glade (file).

However, if you are not comfortable in dig into it, just forget it XD

Reply:
Still too shady :(

I'm sticking with my implementation
It's already clean and eliminated code duplication from canvas_properties.cpp

You can enhance it in the future after merging

Reply:
Ok :) Sorry I couldn't explain it better to you :(